### PR TITLE
KAFKA-6937: In-sync replica delayed during fetch if replica throttle …

### PR DESCRIPTION
**Pull request by junrao**
…is exceeded

* Added a unit test.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by junrao** [comment]
@benstopford : Does this fix look good to you?

**Review by ijuma — APPROVED** [review]
LGTM, but let's wait for @benstopford to take a look too.

**Comment by lindong28** [comment]
Should we also replace `quota.isThrottled(element.getKey)` with `shouldLeaderThrottle(...)` in `KafkaApis.SelectingIterator`?

Also, it seems that `replicaManager.shouldLeaderThrottle(...)` will still be used to determine the FetchResponse in `DelayedFetch.tryComplete()` and `replicaManager.readFromLocalLog()` even if the request is from client. Should we use the replica quota only if the request is from a follower?

**Comment by benstopford** [comment]
@lindong28 
- The replication quota is set to UnboundedQuota if the request is from a client in KafkaApis.replicationQuota(...) so that should ensure client requests are not throttled. 
- The SelectingIterator does not look correct. It appears to erroneously increase the quota metric if the partition is throttled but in sync, which would throw off the leader quota for any out of sync replicas being throttled on the same broker. The same is true for the isThrottled() call in ReplicaFetchThread.processPartitionData (for the follower quota).

@junrao logic/test in your change look good but I think the above point needs addressing. I can take a look at it on tuesday if that helps.

**Comment by junrao** [comment]
@benstopford : I think the KafkaApis.SelectingIterator logic is by design as described in KAFKA-4313. Once a replica is in-sync, we don't want it to be throttled, but we still want to account its traffic towards the replica quota so that the total replication traffic (in sync and not in-sync) doesn't exceed the specified replica quota.

**Comment by benstopford** [comment]
@junrao Ah, OK - is see what you did here (it confused me because I think you took over my PR or something, in that the code looks like I wrote it, but the checkin is yours). 

It might be worth adding a comment to ReplicaFetchThread.processPartitionData + SelectingIterator saying something like "update throttled replicas regardless of whether they are in sync to ensure total replication traffic doesn't exceed quota" as it's pretty subtle, even if you read the jira.

**Comment by lindong28** [comment]
@junrao @benstopford Ah I see. Thanks for the explanation!

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -666,6 +666,8 @@ class KafkaApis(val requestChannel: RequestChannel,
     override def remove() = throw new UnsupportedOperationException()
   }
 
+  // Traffic from both in-sync and out of sync replicas are accounted for in replication quota to ensure total replication
+  // traffic doesn't exceed quota.
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:670]: We don't guarantee this, do we? It seems like we only guarantee that we won't exceed the quota if ISR traffic is below the quota.

Comment (junrao) [reply core/src/main/scala/kafka/server/KafkaApis.scala:670]: Right, we guarantee that if the quota is set high enough.

**Pull request merged**